src/Lumper.Lib/BSP/Struct/PakFileEntry.cs

```diff
@@ -7,50 +7,98 @@ namespace Lumper.Lib.BSP.Struct;
 using Newtonsoft.Json;
 using SharpCompress.Archives.Zip;
 
-public class PakFileEntry
+/// <summary>
+/// A pakfile entry derived from either a ZipArchiveEntry or just a stream.
+/// We assume one is non-null throughout.
+/// </summary>
+public sealed class PakfileEntry : IDisposable
 {
-    public PakFileEntry(ZipArchiveEntry entry)
+    public PakfileEntry(ZipArchiveEntry zipEntry)
     {
-        _entry = entry;
-        Key = entry.Key;
+        ZipEntry = zipEntry;
+        Key = zipEntry.Key ?? throw new InvalidDataException("Pakfile contains an item without a key");
     }
 
-    public PakFileEntry(string key, Stream stream)
+    public PakfileEntry(string key, Stream stream)
     {
         Key = key;
-        _dataStream = stream;
+
+        stream.Seek(0, SeekOrigin.Begin);
+        var mem = new MemoryStream();
+        stream.CopyTo(mem);
+        _buffer = mem.ToArray();
```

Comment:
- stream might not be seekable
- if its a file you don't necessarily want to go to the beginning anyway (but you would need a length)
- using var mem
- copying into stream *and* array is redundant

Reply:
Okay, I'm going to refactor this file quite a bit so take a look after next force-push, I recently realised I need a static lock around Zip.OpenEntryStream so needs some refactoring.

I need a buffer on this class so I can issue multiple read-only streams to the same underlying buffer. I can try to switch this from `ToArray` to `GetBuffer` though to avoid a copy, and you're right that I'm not disposing of the wrapper stream properly.

Take another look at this class as a whole after next push and let me know what you think.